### PR TITLE
Add section on storage to the architecture document

### PR DESCRIPTION
**Pull request by markdryan**
This commit adds a section to the architecture document that describes,
at a high level, how storage works in CIAO.

Fixes: https://github.com/ciao-project/ciao-project.github.io/issues/15

Signed-off-by: Mark Ryan <mark.d.ryan@intel.com>

### CODE REVIEW

File: ciao/arch/architecture.md [hunk 1]
```diff
@@ -423,3 +423,88 @@ various network constructs such as subnets and interfaces are assigned names
...
+read and write to their rootfs, there is a downside.  Any data written to a
+local volume does not survive migration.  The rootfs of a container is reset to
+its default state when it is migrated from one node to another.  It is however
+possible to create a data volume inside the ceph cluster and attach this volume
+to a container.  Any data written to this volume by the container will survive
+migration.
+
+### Images
+
+The controller implements an image service.  This service manages a set of of
```
Comment (rbradford) [inline ciao/arch/architecture.md]: of of

File: ciao/arch/architecture.md [hunk 2]
```diff
@@ -423,3 +423,88 @@ various network constructs such as subnets and interfaces are assigned names
...
+private images, i.e., images that are associated with a given tenant and can
+only be used to create root file systems for instances belonging to the same
+tenant.  Administrators can create public images, that that can be used by all
+users to create VM instances in any tenants they like.
+
+When a new image is uploaded to the image service it is converted into a raw
+format and stored in a new RBD volume in the ceph cluster.  An RBD snapshot is
+then taken of this new volume and the snapshot is protected.  It is this
+protected RBD snapshot that is used as the backing image for VM instances.  The
+image will be assigned a UUID when created.  It is also possible for users to
```
Comment (rbradford) [inline ciao/arch/architecture.md]: image name is now required.

File: ciao/arch/architecture.md [hunk 3]
```diff
@@ -423,3 +423,88 @@ various network constructs such as subnets and interfaces are assigned names
...
+specify a friendly name at creation time.
+
+In order to create an instance that uses an image in the image service as the
+backing image for its rootfs, the backing image must be referenced in the workload
```
Comment (rbradford) [inline ciao/arch/architecture.md]: s/must/can be/

You can create the volume from the image by hand and then use that in the workload definition so use the image in the workload is not the only way to achieve this.

File: ciao/arch/architecture.md [hunk 4]
```diff
@@ -423,3 +423,88 @@ various network constructs such as subnets and interfaces are assigned names
...
+definition used to create the instance.  The backing image can be referenced by
+UUID or by its friendly name.  When the controller creates the new VM, it first
+creates its rootfs, by cloning the protected snapshot that represents the backing
```
Comment (rbradford) [inline ciao/arch/architecture.md]: can you make it clear that the clone is shallow with copy-on-write semantics for the blocks that make it up.

File: ciao/arch/architecture.md [hunk 5]
```diff
@@ -423,3 +423,88 @@ various network constructs such as subnets and interfaces are assigned names
...
+workload definition, but doing so will essentially place a restriction of one
+on the number of instances that can be created from that workload definition, as a
+volume can only be attached to one instance at any one time.
+
+Volumes created by a workload definition can be marked as ephemeral.  Ephemeral
+volumes are automatically deleted when the instance to which they are attached
+is itself deleted. Volumes automatically created to serve as the root file
+systems for instances are marked as ephemeral by default.  Non-ephemeral
+volumes, on the other hand, are not deleted along with the instances they were
+original created with.  They allow data to surive the destruction of the
```
Comment (rbradford) [inline ciao/arch/architecture.md]: survive

Comment (rbradford) [inline ciao/arch/architecture.md]: Maybe use deletion rather than destruction as that is the verb we use in the CLI

**Review by rbradford — APPROVED** [review]
Just some minor nits.

**Comment by markdryan** [comment]
@rbradford Thanks for the review.  All comments have been addressed.

**Pull request merged**